### ISSUE TITLE
Error messages lie about dependencies of terms on Let-bound identifiers

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#8153]
#### Version

8.8.1

#### Description of the problem

This code fails, but it should succeed.

```coq
Section foo.
  Variable (x : nat).

  Let y := x.

  Lemma foo : x = 0 + x.
  Proof using Type. clear y. reflexivity. Qed.

  Lemma bar : x = 0 + x.
  Proof.
    clear y.
    rewrite <- foo. (* Error: foo depends on the variable y which is not declared in the context. *)
    reflexivity.
  Qed.
End foo.
```

**Comment by SkySkimmer** [comment]
Duplicate of #8154.

**Comment by JasonGross** [comment]
No, I don't think this is a duplicate, even if the same commit is responsible for both.  In particular, this is very clearly a bug that makes developing some things much more annoying, while #8154 is closer to a disagreement with the current spec.  Also, it is theoretically possible to separate the behavior, and have `Set Proof Using Clear Unused` not clear let-bound definitions, while still fixing this bug.

cc'ing @gares here

**Comment by SkySkimmer** [comment]
I guess if `foo` is opaque it could work, but we still have to check that `y` doesn't appear in its type somehow.

**Comment by JasonGross** [comment]
The issue is that there's a disagreement between different parts of `Proof using` about the status of let binders.  In the example, `Proof using Type` for `foo` apparently allows access to all let binders depending only on variables in the type, under the (correct) presumption that using them doesn't require introducing additional variable dependencies.  However, this logic is backwards for checking the use of such lemmas, where it's incorrect to presume that access to let's is required.  One way to fix this is to give Let's the status of variables for the purposes of Proof using analysis (change the behavior of `Proof using Type`).  Another is to actually track dependency of proofs on Let's (change the behavior of `Qed`, I think)
